var/spack/repos/builtin/packages/wps/package.py

```diff
@@ -41,6 +41,9 @@ class Wps(Package):
     depends_on("wrf")
     depends_on("netcdf-c")
     depends_on("netcdf-fortran")
+    # Dependence on mpi needs to be explicit because concretizers
+    # won't do the right thing and will mix mpi stacks.
```

Comment:
Can you elaborate on this comment? In which case the concretizer mixes different MPIs?

Reply:
Sure. I'm using a spec matrix: 
```  
  specs:
  - matrix:
    - - wps@4.3.1
    - - ^mpich
      - ^openmpi
    - - '%gcc'
      - '%intel'
    - - target=broadwell
  ```
  When I feed this to the original concretizer, it collapses everything to build with openmpi:
```
==> Concretized wps@4.3.1%gcc arch=linux-None-broadwell ^mpich
 -   v3k3wjl  wps@4.3.1%gcc@12.1.0 build_type=serial patches=3f0d999,62c1bcc,e208f47,e86d029 arch=linux-rocky8-broadwell
[+]  6bpdesv      ^jasper@3.0.3%gcc@12.1.0+jpeg~opengl+shared build_type=Release arch=linux-rocky8-broadwell
[+]  gnjtilz      ^libtool@2.4.7%gcc@12.1.0 arch=linux-rocky8-broadwell
[+]  c4txrpb      ^m4@1.4.19%gcc@12.1.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-rocky8-broadwell
[+]  wegvtnf      ^netcdf-c@4.8.1%gcc@12.1.0~dap~fsync~hdf4~jna+mpi~parallel-netcdf+pic+shared patches=de556da arch=linux-rocky8-broadwell
[+]  5dkrthy      ^netcdf-fortran@4.5.4%gcc@12.1.0~doc+pic+shared arch=linux-rocky8-broadwell
[-]  ay462u3      ^tcsh@6.24.00%gcc@12.1.0 patches=3a4e60f arch=linux-rocky8-broadwell
[+]  thkbb4r          ^ncurses@6.2%gcc@12.1.0~symlinks+termlib abi=none arch=linux-rocky8-broadwell
[-]  u7wys7f      ^time@1.9%gcc@12.1.0 arch=linux-rocky8-broadwell
[+]  hkdvgu7      ^wrf@4.3.3%gcc@12.1.0+pnetcdf build_type=dmpar compile_type=em_real nesting=basic patches=01c0f71,68548f6,b3f063c,e07c39c,e4971f6,e6f3db3,f3dd50d,fa78635 arch=linux-rocky8-broadwell

==> Concretized wps@4.3.1%intel arch=linux-None-broadwell ^mpich
 -   zz5gyns  wps@4.3.1%intel@2021.4.0 build_type=serial patches=3f0d999,62c1bcc,e208f47,e86d029 arch=linux-rocky8-broadwell
[+]  xadohtm      ^jasper@3.0.3%intel@2021.4.0+jpeg~opengl+shared build_type=Release arch=linux-rocky8-broadwell
[+]  fqw2cru      ^libtool@2.4.7%intel@2021.4.0 arch=linux-rocky8-broadwell
[+]  rafopof      ^m4@1.4.19%intel@2021.4.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-rocky8-broadwell
[+]  mt4kcfj      ^netcdf-c@4.8.1%intel@2021.4.0~dap~fsync~hdf4~jna+mpi~parallel-netcdf+pic+shared patches=de556da arch=linux-rocky8-broadwell
[+]  kesnuau      ^netcdf-fortran@4.5.4%intel@2021.4.0~doc+pic+shared arch=linux-rocky8-broadwell
[-]  tazx6it      ^tcsh@6.24.00%intel@2021.4.0 patches=3a4e60f arch=linux-rocky8-broadwell
[+]  bwwkhdv          ^ncurses@6.2%intel@2021.4.0~symlinks+termlib abi=none arch=linux-rocky8-broadwell
[-]  cgq2gfv      ^time@1.9%intel@2021.4.0 arch=linux-rocky8-broadwell
[+]  h4344g4      ^wrf@4.3.3%intel@2021.4.0+pnetcdf build_type=dmpar compile_type=em_real nesting=basic patches=01c0f71,68548f6,b3f063c,e07c39c,e4971f6,e6f3db3,f3dd50d,fa78635 arch=linux-rocky8-broadwell

==> Concretized wps@4.3.1%gcc arch=linux-None-broadwell ^openmpi
 -   v3k3wjl  wps@4.3.1%gcc@12.1.0 build_type=serial patches=3f0d999,62c1bcc,e208f47,e86d029 arch=linux-rocky8-broadwell
[+]  6bpdesv      ^jasper@3.0.3%gcc@12.1.0+jpeg~opengl+shared build_type=Release arch=linux-rocky8-broadwell
[+]  gnjtilz      ^libtool@2.4.7%gcc@12.1.0 arch=linux-rocky8-broadwell
[+]  c4txrpb      ^m4@1.4.19%gcc@12.1.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-rocky8-broadwell
[+]  wegvtnf      ^netcdf-c@4.8.1%gcc@12.1.0~dap~fsync~hdf4~jna+mpi~parallel-netcdf+pic+shared patches=de556da arch=linux-rocky8-broadwell
[+]  5dkrthy      ^netcdf-fortran@4.5.4%gcc@12.1.0~doc+pic+shared arch=linux-rocky8-broadwell
[-]  ay462u3      ^tcsh@6.24.00%gcc@12.1.0 patches=3a4e60f arch=linux-rocky8-broadwell
[+]  thkbb4r          ^ncurses@6.2%gcc@12.1.0~symlinks+termlib abi=none arch=linux-rocky8-broadwell
[-]  u7wys7f      ^time@1.9%gcc@12.1.0 arch=linux-rocky8-broadwell
[+]  hkdvgu7      ^wrf@4.3.3%gcc@12.1.0+pnetcdf build_type=dmpar compile_type=em_real nesting=basic patches=01c0f71,68548f6,b3f063c,e07c39c,e4971f6,e6f3db3,f3dd50d,fa78635 arch=linux-rocky8-broadwell

==> Concretized wps@4.3.1%intel arch=linux-None-broadwell ^openmpi
 -   zz5gyns  wps@4.3.1%intel@2021.4.0 build_type=serial patches=3f0d999,62c1bcc,e208f47,e86d029 arch=linux-rocky8-broadwell
[+]  xadohtm      ^jasper@3.0.3%intel@2021.4.0+jpeg~opengl+shared build_type=Release arch=linux-rocky8-broadwell
[+]  fqw2cru      ^libtool@2.4.7%intel@2021.4.0 arch=linux-rocky8-broadwell
[+]  rafopof      ^m4@1.4.19%intel@2021.4.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-rocky8-broadwell
[+]  mt4kcfj      ^netcdf-c@4.8.1%intel@2021.4.0~dap~fsync~hdf4~jna+mpi~parallel-netcdf+pic+shared patches=de556da arch=linux-rocky8-broadwell
[+]  kesnuau      ^netcdf-fortran@4.5.4%intel@2021.4.0~doc+pic+shared arch=linux-rocky8-broadwell
[-]  tazx6it      ^tcsh@6.24.00%intel@2021.4.0 patches=3a4e60f arch=linux-rocky8-broadwell
[+]  bwwkhdv          ^ncurses@6.2%intel@2021.4.0~symlinks+termlib abi=none arch=linux-rocky8-broadwell
[-]  cgq2gfv      ^time@1.9%intel@2021.4.0 arch=linux-rocky8-broadwell
[+]  h4344g4      ^wrf@4.3.3%intel@2021.4.0+pnetcdf build_type=dmpar compile_type=em_real nesting=basic patches=01c0f71,68548f6,b3f063c,e07c39c,e4971f6,e6f3db3,f3dd50d,fa78635 arch=linux-rocky8-broadwell
```
The hashes for what should be the mpich builds match the hashes for the openmpi builds. 

With clingo, it doesn't concretize at all:
```
==> Starting concretization pool with 4 processes
==> Error: errors occurred during concretization of the environment:
'mpich' is not a valid dependency for any package in the DAG
'mpich' is not a valid dependency for any package in the DAG
'openmpi' is not a valid dependency for any package in the DAG
'openmpi' is not a valid dependency for any package in the DAG
```

I'm building up my installation in layers, e.g. a layer for mpi, a layer for hdf5, a layer for netcdf, a layer for wrf, and finally this layer for wps, and declaring builds of dependencies from lower layers as externals. 

When I add this dependency of wps on mpi, it magically works with both concretizers.

Reply:
> When I add this dependency of wps on mpi, it magically works with both concretizers.

Can you remove the comment then? The point is that a spec matrix is built upon the metadata in `package.py` - not the other way around. 

The other question is: does `wps` depend on `mpi` at build time?

Reply:
I removed the comment. 

Presumably yes, it depends on mpi at build time because it needs to use the (correct) mpi compiler wrappers.